OWML.ModHelper.Input/ModInputHandler.cs

```diff
@@ -0,0 +1,349 @@
+using System.Security.Policy;
+
+namespace OWML.ModHelper.Input
+{
+    public class ModInputHandler : IModInputHandler
+    {
+        private const float Cooldown = 0.05f;
+        private const float TapDuration = 0.1f;
+        private const int MinUsefulKey = 8;
+        private const int MaxUsefulKey = 350;
+        private const int MaxComboLength = 7;
+        private const BindingFlags NonPublic = BindingFlags.NonPublic | BindingFlags.Instance;
+
+        internal static ModInputHandler Instance { get; private set; }
+
+        private HashSet<IModInputCombination> _singlesPressed = new HashSet<IModInputCombination>();
+        private Dictionary<long, IModInputCombination> _comboRegistry = new Dictionary<long, IModInputCombination>();
+        private HashSet<InputCommand> _gameBindingRegistry = new HashSet<InputCommand>();
+        private HashSet<IModInputCombination> _toResetOnNextFrame = new HashSet<IModInputCombination>();
+        private float[] _timeout = new float[MaxUsefulKey];
+        private int[] _gameBindingCounter = new int[MaxUsefulKey];
+        private IModInputCombination _currentCombination;
+        private int _lastSingleUpdate;
+        private int _lastCombinationUpdate;
+        private readonly IModLogger _logger;
+        private readonly IModConsole _console;
+
+        internal bool ShouldIgnore(KeyCode code)
+        {
+            UpdateCurrentCombination();
+            return _currentCombination != null && Time.realtimeSinceStartup - _timeout[(int)code] < Cooldown;
+        }
+
+        public ModInputHandler(IModLogger logger, IModConsole console, IHarmonyHelper patcher)
+        {
+            _console = console;
+            _logger = logger;
+            patcher.AddPrefix<SingleAxisCommand>("ClearBinding", typeof(InputInterceptor), nameof(InputInterceptor.SingleAxisRemovePre));
+            patcher.AddPrefix<DoubleAxisCommand>("ClearBinding", typeof(InputInterceptor), nameof(InputInterceptor.DoubleAxisRemovePre));
+            patcher.AddPostfix<SingleAxisCommand>("Update", typeof(InputInterceptor), nameof(InputInterceptor.SingleAxisUpdatePost));
+            patcher.AddPostfix<DoubleAxisCommand>("Update", typeof(InputInterceptor), nameof(InputInterceptor.DoubleAxisUpdatePost));
+            Instance = this;
+        }
+
+        private long HashFromKeyboard()
+        {
+            long hash = 0;
+            var keysCount = 0;
+            var keys = new int[MaxComboLength];
+            var countdownTrigger = true;
+            for (var code = MinUsefulKey; code < MaxUsefulKey; code++)
+            {
+                if (!(Enum.IsDefined(typeof(KeyCode), (KeyCode)code) && UnityEngine.Input.GetKey((KeyCode)code)))
+                {
+                    continue;
+                }
+                keys[keysCount] = code;
+                keysCount++;
+                if (keysCount > MaxComboLength)
+                {
+                    return (int)RegistrationCode.CombinationTooLong;
```

Comment:
return null instead?
return type could be `long?` (nullable long)

Reply:
the reason why I made it that way was to make it consistant with the way conversion from string to hash is handled, but there isn't much point in doing that